<a href="https://colab.research.google.com/github/aghakaleem/Classification-on-Diabetes-Dataset-/blob/main/Classification_on_Diabetes_Dataset_Agha_Kaleemullah.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
pip install watermark

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 11.5 MB/s eta 0:00:00


In [2]:
%load_ext watermark
%watermark -a "Agha Kaleemullah Khan" -u -d -p numpy,pandas,matplotlib,sklearn

Author: Agha Kaleemullah Khan

Last updated: 2024-02-18

numpy     : 1.25.2
pandas    : 1.5.3
matplotlib: 3.7.1
sklearn   : 1.2.2



# **Perceptron Model**

## Loading Dataset

In [5]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


In [11]:
!ls /content/drive/MyDrive/Colab\ Notebooks/Dataset/

diabetes.csv


In [19]:
import pandas as pd
df=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Dataset/diabetes.csv",
                 encoding='utf-8')

df.head()

col2  col2.1  col3  col4  col5  col6   col7  col8  y
0     6     148    72    35     0  33.6  0.627    50  1
1     1      85    66    29     0  26.6  0.351    31  0
2     8     183    64     0     0  23.3  0.672    32  1
3     1      89    66    23    94  28.1  0.167    21  0
4     0     137    40    35   168  43.1  2.288    33  1

In [20]:
df.shape

(768, 9)

### Extracting feature columns (X) and target column (y)

In [22]:

X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

### Splitting the data into training and testing sets

In [38]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Perceptron

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y)

In [39]:
X_train

array([[  7.   , 184.   ,  84.   , ...,  35.5  ,   0.355,  41.   ],
       [  6.   ,  85.   ,  78.   , ...,  31.2  ,   0.382,  42.   ],
       [  2.   , 106.   ,  64.   , ...,  30.5  ,   1.4  ,  34.   ],
       ...,
       [  1.   ,  97.   ,  70.   , ...,  38.1  ,   0.218,  30.   ],
       [  6.   , 190.   ,  92.   , ...,  35.5  ,   0.278,  66.   ],
       [  4.   , 144.   ,  58.   , ...,  29.5  ,   0.287,  37.   ]])

In [40]:
import numpy as np

print('Labels count in y:', np.bincount(y))
print('Labels count in y_train:', np.bincount(y_train))
print('Labels count in y_test:', np.bincount(y_test))

Labels count in y: [500 268]
Labels count in y_train: [350 187]
Labels count in y_test: [150  81]


### Standardizing the features:

In [41]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [42]:
X_train_std

array([[ 0.93138344,  2.0179454 ,  0.78066953, ...,  0.43148259,
        -0.37477883,  0.63212912],
       [ 0.63260632, -1.14861888,  0.46538785, ..., -0.1198324 ,
        -0.29416766,  0.71699246],
       [-0.56250219, -0.47692343, -0.2702694 , ..., -0.20958135,
         2.74517192,  0.03808578],
       ...,
       [-0.86127931, -0.76479291,  0.04501228, ...,  0.76483585,
        -0.78380586, -0.30136756],
       [ 0.63260632,  2.20985838,  1.2010451 , ...,  0.43148259,
        -0.60466993,  2.75371249],
       [ 0.03505207,  0.73852549, -0.58555107, ..., -0.33779414,
        -0.57779954,  0.29267578]])

## Training the perceptron via scikit-learn

In [43]:
perceptron_model = Perceptron(eta0=0.1, random_state=1)
perceptron_model.fit(X_train_std, y_train)

Perceptron(eta0=0.1, random_state=1)

## Predicting Y values

In [44]:
y_pred = perceptron_model.predict(X_test_std)
print('Misclassified examples: %d' % (y_test != y_pred).sum())

Misclassified examples: 71


## Accuracy Estimation

In [45]:
from sklearn.metrics import accuracy_score

print('Accuracy: %.3f' % accuracy_score(y_test, y_pred))

Accuracy: 0.693


### Confusion Matrix


In [46]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[104,  46],
       [ 25,  56]])

### Classification Report

In [50]:
from sklearn.metrics import classification_report
target_names = ['0', '1']
print(classification_report(y_test, y_pred, target_names=target_names))

              precision    recall  f1-score   support

           0       0.81      0.69      0.75       150
           1       0.55      0.69      0.61        81

    accuracy                           0.69       231
   macro avg       0.68      0.69      0.68       231
weighted avg       0.72      0.69      0.70       231



# **Adaline Model**

### Split the data into training and testing sets

In [84]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

### Standardizing the features

In [85]:
scaler = StandardScaler()
X_train_std = scaler.fit_transform(X_train)
X_test_std = scaler.transform(X_test)

## Creating the model and fitting

In [86]:
from sklearn.linear_model import SGDClassifier

adaline = SGDClassifier(loss='perceptron', learning_rate='constant', eta0=0.01, max_iter=1000, random_state=42)

adaline.fit(X_train_std, y_train)

SGDClassifier(eta0=0.01, learning_rate='constant', loss='perceptron',
              random_state=42)

### Predicting on the Test Set

In [87]:
predictions = adaline.predict(X_test_std)

### Accuracy Estimation

In [88]:
accuracy = accuracy_score(y_test, predictions)
print(f'Accuracy: {accuracy}')

Accuracy: 0.7012987012987013


### Confusion Matrix

In [89]:
confusion_matrix(y_test, predictions)

array([[73, 27],
       [19, 35]])

### Classification Report

In [90]:

print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.79      0.73      0.76       100
           1       0.56      0.65      0.60        54

    accuracy                           0.70       154
   macro avg       0.68      0.69      0.68       154
weighted avg       0.71      0.70      0.71       154



# **Logistic Regression Model**

### Split the data into training and testing sets

In [91]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

### Standardizing the features

In [92]:
scaler = StandardScaler()
X_train_std = scaler.fit_transform(X_train)
X_test_std = scaler.transform(X_test)

## Creating the model and fitting

In [93]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(C=100.0, random_state=42, solver='lbfgs', multi_class='ovr')
lr.fit(X_train_std, y_train)

LogisticRegression(C=100.0, multi_class='ovr', random_state=42)

### Predicting on the Test Set

In [94]:
predictions = lr.predict(X_test_std)

### Accuracy Estimation

In [95]:
accuracy = lr.score(X_test_std, y_test)
print("Accuracy:", accuracy)

Accuracy: 0.7142857142857143


### Confusion Matrix

In [96]:
confusion_matrix(y_test, predictions)

array([[82, 18],
       [26, 28]])

### Classification Report

In [97]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.76      0.82      0.79       100
           1       0.61      0.52      0.56        54

    accuracy                           0.71       154
   macro avg       0.68      0.67      0.67       154
weighted avg       0.71      0.71      0.71       154



# Which algorithm gives best accuracy?

### Logistic Regression seems to be giving best accuracy among the other algorithms namely Perceptron and Adaline.


##### Accuracy of Logistric regression is 71.4% compared to Adaline (70.1%) and Perceptron (69.3%). More over, Recall and Precision also seems to be best in Logistic Regression.

